packages/expo-modules-core/ios/JSI/EXJSIConversions.mm

```diff
@@ -26,7 +27,18 @@
 
 jsi::String convertNSStringToJSIString(jsi::Runtime &runtime, NSString *value)
 {
-  return jsi::String::createFromUtf8(runtime, [value UTF8String] ?: "");
```

Comment:
`createFromUtf8` has the other variant that allows passing length

```cpp
  /// Create a JS string from utf8-encoded octets.  The string data is
  /// transformed and copied.  Results are undefined if \p utf8 contains invalid
  /// code points.
  static String
  createFromUtf8(Runtime& runtime, const uint8_t* utf8, size_t length) {
    return runtime.createStringFromUtf8(utf8, length);
  }
```

wonder if that works
```objcpp
NSData *data = [value dataUsingEncoding:NSUTF8StringEncoding];
return jsi::String::createFromUtf8(runtime, reinterpret_cast<const uint8_t*>(data.bytes), data.length);
```

Reply:
We tried that, unfortunately, it wasn't working. The string was truncated by Hermes. I think the JS can only interpret the `0 byte` correctly when it's encoded using utf16. In that case, the `0 byte` is replaced by two `0 bytes`.